### ISSUE TITLE
[Chrome] Opening a link from bookmarks toolbar, after not opening a new tab for a while, opens the link at the end

### ISSUE THREAD

**Issue by PuppetMaster000** [autonome/Always-Right#34]
Hoping the extension is still supported. Starting from about 2 months ago roughly, I believe the new Chrome update broke the extension. Whenever I would open a new link from the bookmarks toolbar (using the middle mouse button) but ONLY after not opening anything for about a minute, the first new tab would always go all the way right. Then on the second click, it would go back to normal and open to the right.

It only happens when no new tabs have been opened for a minute, easiest way to test is just to not touch anything, leave the browser running, then after a minute use the scroll wheel (middle mouse button) to open a new link from the bookmarks toolbar.
It opens all the way on the right, but then you can click the exact same link again and it would open to the right as it is suppose to.

The weird part is, you can then keep opening links from the toolbar normally for about 30 seconds, but the moment you don't touch the browser for a minute, it is like the extension "goes to sleep" and needs to wake up again. I should also mention this only happens with the bookmarks toolbar, normal links from a site seem to act as they should.

**Comment by autonome** [comment]
Hi, thanks for filing!

I can reproduce this in latest Chrome by emulating middle mouse button via cmd+click on a bookmark on the bookmarks bar.

Confirmed that a little wait is required to reproduce.

I'll take a look and see what's happening.

**Comment by PuppetMaster000** [comment]
I can't believe I waited this long to report it just because I thought it would never be seen...
You are awesome, thank you!

**Comment by autonome** [comment]
A bit more testing, and this appears to be about tabs.onCreated not firing after a minimum of 25 seconds of delay.

And confirmed that you can reproduce just by cmd+t to open a new tab. The bookmarks bar and middle-click are not relevant. Appears not to matter what the source of the event is.

**Comment by PuppetMaster000** [comment]
Thanks for looking into it so quickly. Yeah, that sounds about right. I personally only use the middle mouse button (scroll wheel) to open tabs, which is why I referenced it, but I was sure it applied to other ways too. I just wasn't familiar with all of them.
In any case, do you believe this is something that can be fixed? And if so, is it something we should do from user side or should we just wait for the update?

**Comment by autonome** [comment]
Ok I confirmed with a minimal extension in a clean profile that it is likely not solely related to tab events, and could be a bug in the extension.

Extension code:

`chrome.tabs.onCreated(() => console.log('🤪'));`

Fires immediately upon opening a new tab, even after the 25s delay.

**Comment by PuppetMaster000** [comment]
I suppose that is good news, better if it is in the extension than Chrome itself right? I would assume the extension is easier to control compared to the browser, so hopefully it can be fixed.

**Comment by autonome** [comment]
Yes and no.

I'm experiencing the observer/observed problem at the moment 😂

https://bsky.app/profile/burrito.space/post/3kqpght7j7c2y

**Comment by autonome** [comment]
This is likely the Chromium bug: https://issues.chromium.org/issues/40805401

I'm testing a workaround: https://stackoverflow.com/a/66618269/4096929

**Comment by autonome** [comment]
Ok looks like it's working so far.

I'll ship an update so you can try.

**Comment by autonome** [comment]
Hrm, nope. Just tested opening external link after 25s delay and still opened at end.

**Comment by PuppetMaster000** [comment]
Damn, this bug seems so simple at first, but seems like a nightmare to figure out actually. Thank you for the hard work, hope you can find a fix.

**Comment by autonome** [comment]
I might just ship with this attempted mitigation and call it a day, hope it happens at least a bit less.

This is clearly a bug in Chrome's implementation of mv3 extensions... which they do not consider a bug.

All the other workarounds add both complexity and unnecessary resource consumption, and put the extension at risk of getting flagged and booted.

**Comment by PuppetMaster000** [comment]
I checked out the extension called "Open Tabs Next to Current" from the Chrome store, which seems to not have the issue, but it acts differently when opening multiple tabs. I much preferred yours because when opening more tabs from the current tab location, it would open to the right in order clicked (current tab, 1st, 2nd, 3rd clicked), but their extension always opens everything to the right of the current tab you are on only (current, 3rd, 2nd, 1st clicked) if that makes any sense. Any tab opened is forced exactly to the right of the current one.
I am only mentioning this if you wish to (although not sure if you can even) check that extension out and see why that one doesn't have the 25s delay thing. Here is a direct link to it in Chrome store:
https://chromewebstore.google.com/detail/open-tabs-next-to-current/gmpnnmonpnnmnhpdldahlekfofigiffh

Again, sorry if what I am saying is ignorant, I have no idea how any of this works, just sharing what I have noticed.

**Comment by autonome** [comment]
Super helpful, thanks! I'll if the source is available, see what they're doing differently.

**Comment by PuppetMaster000** [comment]
Glad I could help ^^
Sorry I don't know what mv2 and mv3 means 😢 Is it not possible to make yours mv2 as well without issues?

**Comment by autonome** [comment]
It's the manifest version - and is basically like "major versions" of the browser extension framework itself.

Google developed and pushed manifest v3, and is no longer accepting v2 extensions in the store. But it turns out their own implementation of v3 is basically a bug-infested horror show.

But now that we're having this conversation, I can't remember if the version in the Chrome store is the v3 migration or not...

If it is, there's no way to go back.

If it's not, then I've been chasing the bug in the wrong codebase. Which isn't a loss - I know now that mv3 will have this problem until the Chrome bug is fixed.

I'll check the source of the store version, see what's what.

**Comment by autonome** [comment]
Hrm, looks like v2.0.0 in the store is the mv3 version.

I'm sorry, but unfortunately I think we'll just have to wait until Chrome fixes the underlying issue. I can't put any more time into working around the fact that they shipped broken stuff.

Maybe try switching to Firefox, where they implemented Chrome's own extensions API correctly 😅

**Comment by PuppetMaster000** [comment]
I see, thank you for clarifying! I actually use both Chrome and Firefox haha, got the extension on both. Shame, I appreciate the effort. I understand its not reasonable to keep trying to work on this if it is mostly on Chrome. I guess if you can, report it to whoever or wherever Chrome gets info on bugs and hope for the best. I will wait for your update to come out and can report if it made things better at least. Until then, I suppose I will stick with that alternative extension. Maybe shoot them a message and see if they are able to make it closer to yours in behavior haha.

Again, thanks for all the effort these past few days!

**Comment by autonome** [comment]
No worries, thanks for reporting it.

I did some more testing, but this time against Chrome 124 (stable) vs Chrome 126 (nightly), and in early results appears that the workaround does reduce or completely remove incidence of the issue. I can't reproduce it at all when the workaround is in place (so far). I can reproduce 100% of the time without the workaround.

So I'll ship an updated version w/ the workaround and hopefully the problem is resolved by the time 126 hits stable channel. Or maybe sooner - I didn't test 125.

**Comment by autonome** [comment]
submitted to chrome store for review

**Comment by PuppetMaster000** [comment]
Hey just wanted to add that there is unfortunately another issue after the update. The original issue where it would jump is no longer there, but now, when reopening the browser on Chrome, it takes the tab that is furthest on the right and moves it next to the one you were on it seems.
So after opening the browser a few times, it would move several tabs around and mess things up. It would go from:
1,2,3,4,5 to 1,5,2,3,4 then when you open another time 1,4,5,2,3 and so on. I am not 100% certain this is the exact explanation of the issue, but in any case, tabs from the end are definitively moving.

**Comment by autonome** [comment]
Can you file a new issue? Please include whether it's consistent, or sporadic.

**Comment by baohaojun** [comment]
Hi, I seet to have fixed this issue with help of AI, in claude code, I asked it to fix it, and here's AI's output, I have tested, it works.

❯ this extension sometimes works, but sometimes does not work on a newly opened tab, the 2nd new tab will open to the far right end, why?  help me fix it.                                                                                                                          
⏺ Read 2 files (ctrl+o to expand)

⏺ Let me read the memory file and understand the code flow.

⏺ Read 1 file (ctrl+o to expand)

⏺ I can see the bug. Here's the race condition:

  1. Open new tab 1: onCreated fires → makeRight starts with activeTabCache = X (index 0)
  2. onActivated fires for new tab 1 → activeTabCache = {id: newTab1, index: 3} (at the end, before move)
  3. makeRight moves new tab 1 to index 1. Tab strip: [X(0), newTab1(1), A(2), B(3)]
  4. Open new tab 2: onCreated fires → makeRight starts with activeTabCache = {id: newTab1, index: 3} ← STALE!
  5. targetIndex = 3 + 1 = 4, but new tab 2 is already at index 4, so newTab.index == targetIndex → returns early without moving

  The fix: capture the activeTabId synchronously upfront, then after fetching the window, look up the tab's current (fresh) index from the   live win.tabs list.

⏺ Update(service_worker.js)
  ⎿  Added 18 lines, removed 11 lines
      42    // No active tab yet — still in session restore or pre-interaction.
      43    if (!activeTabCache) return;
      44   
      45 -  const activeTab = activeTabCache;                                                                                      
      45 +  // Capture the active tab ID synchronously before any await, so we can look up                                         
      46 +  // its fresh position later (activeTabCache.index goes stale when tabs are moved).                                     
      47 +  const activeTabId = activeTabCache.id;                                                                                 
      48   
      49    // The new tab either dragged to new window or something went wrong.
      50 -  if (newTab.windowId != activeTab.windowId) {                                                                           
      50 +  if (newTab.windowId != activeTabCache.windowId) {                                                                      
      51      return;
      52    }
      53   
      54 +  // We need current window for a few things required for correct tab placement.                                         
      55 +  const win = await api.windows.getCurrent({ populate: true });                                                          
      56 +                                                                                                                         
      57 +  // Maybe is a restored tab, or another add-on, or something else is wonky.                                             
      58 +  if (newTab.index < win.tabs.length - 1 || newTab.index > win.tabs.length - 1) {                                        
      59 +    return;                                                                                                              
      60 +  }                                                                                                                      
      61 +                                                                                                                         
      62 +  // Get the active tab's current position from the live window snapshot.                                                
      63 +  // This avoids stale index: onActivated captures index before makeRight moves                                          
      64 +  // the tab, so a 2nd new tab would see the wrong targetIndex from cache.                                               
      65 +  const activeTab = win.tabs.find(t => t.id === activeTabId);                                                            
      66 +  if (!activeTab) return;                                                                                                
      67 +                                                                                                                         
      68    // To the right
      69    let targetIndex = activeTab.index + 1;
      70   
     ...
      58      return;
      59    }
      60   
      61 -  // We need current window for a few things required for correct tab placement.                                         
      62 -  // And apparently tab references go STALE. Dammit.                                                                     
      63 -  const win = await api.windows.getCurrent({ populate: true });                                                          
      64 -                                                                                                                         
      65 -  // Maybe is a restored tab, or another add-on, or something else is wonky.                                             
      66 -  if (newTab.index < win.tabs.length - 1 || newTab.index > win.tabs.length - 1) {                                        
      67 -    return;                                                                                                              
      68 -  }                                                                                                                      
      69 -                                                                                                                         
      61    // If the active tab is pinned, we have to set the target index
      62    // to that of the first non-pinned tab.
      63    if (activeTab.pinned) {

⏺ The root cause and fix:

  Bug: activeTabCache stores the tab's index at the moment onActivated fires — which is before makeRight moves that tab. So when you open
  a 2nd new tab, activeTabCache.index is stale (e.g., 3 instead of the actual 1), making targetIndex = 4, which matches the new tab's
  position at the end → early return, no move.

  Fix: Instead of using const activeTab = activeTabCache (stale copy), the code now:
  1. Captures activeTabId synchronously upfront (just the ID, which never changes)
  2. Moves the windows.getCurrent() call earlier (before computing targetIndex)
  3. Looks up the active tab in win.tabs by ID — this gives the current position from the live window snapshot, not the stale cached one
